docs/adr/0020-application-architecture.md

```diff
@@ -0,0 +1,123 @@
+# Use a documented Application Architecture for new back-end services
+
+- Deciders: Ben Bangert, Ryan Kelly, Vijay Budhram, Jody Heavener
+- Date: 2020-08-12
+
+## Context and Problem Statement
+
+The FxA back-end service stack contains an application architecture that is ad-hoc, not documented, and missing modern features (such as Dependency Injection) which results in the following problems:
+
+- New developers struggle to get up to speed as they must learn the architecture by reading the code as we have no documentation on the application structure, why they're structured the way they are, or how new components should be added to fit in. Each back-end service may vary in its ad-hoc architecture as well.
+- Adding new objects needed in a route handler can be time-consuming as the object must be plumbed through the entire initialization chain vs. more elegant methods like Dependency Injection (DI).
+- Not clear where/how to add new components and takes time to study/understand how things are currently setup in an attempt to mimic the structure for the new component.
+- Time consuming to setup boiler-plate for components, as we have no tooling to work with the current ad-hoc application architectures.
+- Our ad-hoc architecture frequently mixes concerns such as having business logic mixed in with request handling logic, and has other warts from its evolution over time vs. being planned up front.
+- New back-end services evolve differently resulting in more ad-hoc application architectures to learn.
+- Shared components in `fxa-shared` can't rely on basic object lifecycles or setup approaches as they may be used in multiple different ad-hoc application architectures.
```

Comment:
Side note, if there's an opportunity here to re-purpose `fxa-shared` so its not backend-specific, or rename it so it's less generic sounding, I'd like to do this.